language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/ExBudgeting.hs

```diff
@@ -0,0 +1,100 @@
+    , estimateStaticStagedCost
+    , exBudgetCPU
+    , exBudgetMemory
+    , exTallyMemory
+    , exTallyCPU
+    , exBudgetStateBudget
+    , exBudgetStateTally
+    )
+where
+
+
+import           Language.PlutusCore
+import           PlutusPrelude
+
+import           Control.Lens.TH                                         (makeLenses)
+import           Data.HashMap.Monoidal
+import           Language.PlutusCore.Evaluation.Machine.ExMemory
+import           Language.PlutusCore.Evaluation.Machine.GenericSemigroup
+
+data CekBudgetMode =
+      Counting -- ^ For precalculation
+    | Restricting -- ^ For execution, to avoid overruns
+
+data ExBudget = ExBudget { _exBudgetCPU :: ExCPU, _exBudgetMemory :: ExMemory }
+    deriving (Eq, Show, Generic)
+    deriving (Semigroup, Monoid) via (GenericSemigroupMonoid ExBudget)
+
+data ExBudgetState = ExBudgetState
+    { _exBudgetStateTally  :: ExTally -- ^ for counting what cost how much
+    , _exBudgetStateBudget :: ExBudget  -- ^ for making sure we don't spend too much
+    }
+    deriving stock (Eq, Generic, Show)
+
+data ExTally = ExTally
+    { _exTallyCPU    :: MonoidalHashMap (Plain Term) [ExCPU]
+    , _exTallyMemory :: MonoidalHashMap (Plain Term) [ExMemory]
+    }
+    deriving stock (Eq, Generic, Show)
+    deriving (Semigroup, Monoid) via (GenericSemigroupMonoid ExTally)
+
+$(join <$> traverse makeLenses [''ExBudgetState, ''ExBudget, ''ExTally])
+
+-- TODO See language-plutus-core/docs/Constant application.md for how to properly implement this
+estimateStaticStagedCost
+    :: BuiltinName -> [WithMemory Value] -> (ExCPU, ExMemory)
+-- estimateStaticStagedCost AddInteger           [x, y] = (1, 1)
+-- estimateStaticStagedCost SubtractInteger      [x, y] = (1, 1)
+-- estimateStaticStagedCost MultiplyInteger      [x, y] = (1, 1)
+-- estimateStaticStagedCost DivideInteger        [x, y] = (1, 1)
+-- estimateStaticStagedCost QuotientInteger      [x, y] = (1, 1)
+-- estimateStaticStagedCost RemainderInteger     [x, y] = (1, 1)
+-- estimateStaticStagedCost ModInteger           [x, y] = (1, 1)
+-- estimateStaticStagedCost LessThanInteger      [x, y] = (1, 1)
+-- estimateStaticStagedCost LessThanEqInteger    [x, y] = (1, 1)
+-- estimateStaticStagedCost GreaterThanInteger   [x, y] = (1, 1)
+-- estimateStaticStagedCost GreaterThanEqInteger [x, y] = (1, 1)
+-- estimateStaticStagedCost EqInteger            [x, y] = (1, 1)
+-- estimateStaticStagedCost Concatenate          [x, y] = (1, 1)
+-- estimateStaticStagedCost TakeByteString       [x, y] = (1, 1)
+-- estimateStaticStagedCost DropByteString       [x, y] = (1, 1)
+-- estimateStaticStagedCost SHA2                 x      = (1, 1)
+-- estimateStaticStagedCost SHA3                 x      = (1, 1)
+-- estimateStaticStagedCost VerifySignature      [x, y] = (1, 1)
+-- estimateStaticStagedCost EqByteString         [x, y] = (1, 1)
+-- estimateStaticStagedCost LtByteString         [x, y] = (1, 1)
+-- estimateStaticStagedCost GtByteString         [x, y] = (1, 1)
```

Comment:
I don't think the commented-out code helps much.

Reply:
It's mostly a bit less work for whoever implements this. I could delete it.

Reply:
YAGNI. In particular, given our discussion about passing in the the costing functions for builtins along with their interpretations, I doubt that this function is going to survive in its current form.